File: 3sum.rb
# http://programmingpraxis.com/2013/06/18/3sum/

def find_zerosum_terms(array)
  solutions =[]
  array_length = array.length 
  (0...array_length).each do |outer_index|
    (outer_index...array_length).each do |inner_index|
      test_val = -(array[outer_index] + array[inner_index])
      if array.include?(test_val)
        solutions << [array[outer_index],array[inner_index],test_val]
      end  
    end
  end
  solutions.each do |solution|
    solution.sort!
  end
  solutions.uniq
end

test_array = [8, -25, 4, 10, -10, -7, 2, -3]  

p find_zerosum_terms(test_array)

File: 3sum.js
#!/usr/bin/env node
// find three 3 elements which sum to 0

function findZeroSumTerms(list) {
  var solutions = {}
  for (var i in list) {
    for (var j = i; j < list.length; j++) {
      var testValue = -(list[i] + list[j]);
      for (var k in list) {
        if (list[k] === testValue) {
          var values = [list[i],list[j],list[k]].sort(function(a,b) { return a - b});
          var key = values[0] + "," + values[1] + "," + values[2] + ""
          solutions[key] = values
        }
      }
    }
  }  
  var keys = Object.keys(solutions);
  var results = [];
  for (var k in keys) { results.push(solutions[keys[k]]) };
  return results
}

var t = [8,-25,4,10,-10,-7,2,-3];
console.log(findZeroSumTerms(t));
